verifiable-db/src/query/batching/row_chunk/consecutive_rows.rs

```diff
@@ -0,0 +1,1290 @@
+    let first_node_succ_in_range = b.and(smaller_than_max, first.successor_info.is_found);
+    // if first_node_succ_in_range is true, and the successor of the first node was found in the path from
+    // such node to the root of the tree, then the hash of successor node will be placed in
+    // `first.successor_info.hash` by `MerklePathWithNeighborsGadget: therefore, we can check that `second`
+    // is consecutive of `first` by checking that `first.successor_info.hash` is the hash of the second node;
+    // otherwise, we cannot check right now that the 2 nodes are consecutive, we will do it later
+    let check_are_consecutive = b.and(first_node_succ_in_range, first.successor_info.is_in_path);
+    let is_second_node_successor = b.hash_eq(&first.successor_info.hash, &second.end_node_hash);
+    // update are_consecutive as `are_consecutive && is_second_node_successor`` if `check_are_consecutive` is true
+    let new_are_consecutive = b.and(are_consecutive, is_second_node_successor);
+    are_consecutive = BoolTarget::new_unsafe(b.select(
+        check_are_consecutive,
+        new_are_consecutive.target,
+        are_consecutive.target,
+    ));
+    // we now look at the predecessor of second node, matching it with first node in case the
+    // predecessor is found in the path of second node in the tree
+    let second_node_predecessor_value = &second.predecessor_info.value;
+    // ensure that we don't prove nodes outside of the range: the predecessor of the second
+    // node must store a value smaller that `max_query_bound``
+    let smaller_than_max =
+        b.is_less_or_equal_than_u256(second_node_predecessor_value, &max_query_bound);
+    are_consecutive = b.and(are_consecutive, smaller_than_max);
+    // determine whether the predecessor (if any) of the second node stores a value in the query range or not;
+    // note that, since we previously checked that such value is <= max_query_bound,
+    // we only need to check whether this value is not dummy (i.e., if the predecessor exists) and if
+    // such value is >= min_query_bound
+    let bigger_than_min =
+        b.is_less_or_equal_than_u256(min_query_bound, &second_node_predecessor_value);
+    let second_node_pred_in_range = b.and(bigger_than_min, second.predecessor_info.is_found);
+    // if second_node_pred_in_range is true, and the predecessor of the second node was found in the path from
+    // such node to the root of the tree, then the hash of predecessor node will be placed in
+    // `second..predecessor_info.hash` by `MerklePathWithNeighborsGadget: therefore, we can check that `second`
+    // is consecutive of `first` by checking that `second.predecessor_info.hash` is the hash of the first node;
+    // otherwise, we cannot check right now that the 2 nodes are consecutive, and it necessarily means we have
+    // already done it before when checking that the successor of first node was the second node
```

Comment:
When both first and second are in the same tree, one of the conditions must be always true, I argued about this [here](https://www.notion.so/lagrangelabs/Aggregating-Query-Results-with-Individual-Merkle-Paths-10628d1c65a880b1b151d4ac017fa445?pvs=4#10728d1c65a8809cbedccc150417f307).
Indeed, later in this method I compute also a flag `either_is_in_path`, which is true iff one of the 2 conditions is true, and I enforce this flag to be true if the 2 nodes are in the same tree.

Instead, both the conditions might be false if first and second doesn't belong to the same tree: for instance, if `first` is the last node in a rows tree, and `second` is the first node in another rows tree, then `first` has no successor, and `second` has no predecessor, so both the conditions will be false. However, in this case we don't need to bind successor of `first` with `second`, and vice versa, since the node belongs to different rows trees (we are doing other checks to enforce that the nodes are consecutive across rows trees), so it's not an issue to skip both these checks on successor/predecessor hash. Does it clarify the logic here?

Reply:
Got it, thanks !